clippy_lints/src/methods/join_absolute_paths.rs

```diff
@@ -0,0 +1,27 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_ast::ast::LitKind;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_span::symbol::sym::Path;
+
+use super::JOIN_ABSOLUTE_PATHS;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, join_arg: &'tcx Expr<'tcx>) {
+    let ty = cx.typeck_results().expr_ty(expr).peel_refs();
+    if !expr.span.from_expansion() && is_type_diagnostic_item(cx, ty, Path)
+    && let ExprKind::Lit(spanned) = &join_arg.kind && let LitKind::Str(symbol, _) = spanned.node
+    && (symbol.as_str().starts_with('/') || symbol.as_str().starts_with('\\')) {
+      span_lint_and_then(
+        cx,
+        JOIN_ABSOLUTE_PATHS,
+        join_arg.span,
+        "argument to `Path::join` starts with a path separator",
+        |diag| {
+          diag.note("joining a path starting with separator will replace the path instead");
+          diag.help(r#"if this is unintentional, try removing the starting separator"#);
+          diag.help(r#"if this is intentional, try creating a new Path instead"#);
+        },
+      );
+    };
+}
```

Comment:
```suggestion
    if is_type_diagnostic_item(cx, ty, Path)
        && let ExprKind::Lit(spanned) = &join_arg.kind
        && let LitKind::Str(symbol, _) = spanned.node
        && (symbol.as_str().starts_with('/') || symbol.as_str().starts_with('\\'))
    {
        span_lint_and_then(
            cx,
            JOIN_ABSOLUTE_PATHS,
            join_arg.span,
            "argument to `Path::join` starts with a path separator",
            |diag| {
                diag
                      .note("joining a path starting with separator will replace the path instead")
                      .help(r#"if this is unintentional, try removing the starting separator"#)
                      .help(r#"if this is intentional, try creating a new Path instead"#);
            },
        );
    }
```
This formatting is pretty weird, in Clippy we tend to put each `&&` on a new line and the `&&` at the start of the line. iirc there's no consensus in the style team on how this should be formatted, so this is just what we do for now.

Usually, when working with let chains, you want to remove the entire let chain (leaving only the first condition) to see what rustfmt would probably do. This is because, as xFrednet said, rustfmt doesn't format let chains currently, including the entire `then` block.

(This suggestion also fixes the 2 nits ^^)